### PR TITLE
[WIP][BACKLOG-7551] Providing support for multi-valued properties in the "change" event loop

### PR DESCRIPTION
**Pull request by carlosrusso**
@pentaho/millenniumfalcon please review
@swagner8412 Please revise the doclets

### CODE REVIEW

**Comment by lucboudreau** [comment]
Please squash this into as few commits as possible.

**Comment by nantunes** [comment]
Hi @lucboudreau. Are you real or a wingman bot? And why should the commits be squashed?

**Comment by lucboudreau** [comment]
I'm real, according to my last Turing test at least. :) 

Let me throw that question right back at you. Why do you think we need to squash them? I can easily come up with 3 reasons. How many can you find?

**Comment by nantunes** [comment]
I've been getting a lot of comments from your account these past few days, so it was a fair question. :-) It could have been an automatic message in PRs with more than _x_ commits.

I'm really not seeing the reasons, so if you could share yours we could probably reach a better understanding.

**Comment by brosander** [comment]
Base build failed with return code 1

File: package-res/resources/web/pentaho/type/changes/Add.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   * @amd pentaho/type/changes/Add
+   *
+   * @classDesc Describes an operation that adds an element to a list.
+   *
+   * @constructor
+   * @description Creates an instance.
+   *
+   * @param {!pentaho.type.Element} elem - The object to be added to the list.
+   * @param {number} index - The position in the list at which the element should to be inserted.
+   * @param {string} key - The key that should be used for identifying the object to be added.
```
Comment (nantunes) [inline package-res/resources/web/pentaho/type/changes/Add.js]: Isn't this always the same than `elem.key`?

**Comment by dcleao** [comment]
Come on @lucboudreau... This is not a matter of quantity as I'm sure you know. And if you already know you have 3 reasons, why didn't you enumerate them? Would have be nice, for not making us have to guess your game.

Anyway. I don't wanna enter that game, so these, I think, are our main motivations:
1. Maintain the authorship of commits to some extent
2. Maintain readability of changes, by not mixing them when they are very disparate
3. Squash commits as much as possible as long as we don't break the last two rules
4. Group commits by issue

File: package-res/resources/web/pentaho/type/changes/Change.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    get owner() {
+      return this._owner;
+    },
+
+    /**
+     * Assigns a value to this change.
+     *
+     * @method
+     * @abstract
```
Comment (dcleao) [inline package-res/resources/web/pentaho/type/changes/Change.js]: Doesn't this method have parameters or return values?

File: package-res/resources/web/pentaho/type/changes/Change.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    set: null,
+
+    /**
+     * Updates the owning [complex]{@linkplain pentaho.type.Complex} with this change.
+     *
+     * @method
+     * @protected
+     * @abstract
+     */
+    _commit: null
```
Comment (dcleao) [inline package-res/resources/web/pentaho/type/changes/Change.js]: Idem

File: package-res/resources/web/pentaho/type/changes/Change.js (hunk 2 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/Change.js]: The signature of the `set` method depends on the final classes. 
It does not return a value.

File: package-res/resources/web/pentaho/type/changes/Change.js (hunk 3 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/Change.js]: The `_commot` method does not have arguments nor does it return a value.

File: package-res/resources/web/pentaho/type/changes/Changeset.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    get type() {
+      return "changeset";
+    },
+
+    /**
+     * Determines if this changeset contains the specified property.
+     *
+     * @param {nonEmptyString|!pentaho.type.Property.Type} name - The property name or type object.
+     *
+     * @return {boolean} `true` if the property exists and is referenced in this changeset, `false` otherwise.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: suggested wording: 

'true' if the property exists and is referenced in this changeset, 'false' if the property does not exist or is not referenced.

File: package-res/resources/web/pentaho/type/changes/Changeset.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+     *
+     * @type !string[]
+     * @readonly
+     */
+    get propertyNames() {
+      return Object.keys(this._properties);
+    },
+
+    /**
+     * Asserts if this changesets contains any defined changes.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: Asserts if this changeset contains any defined changes.

File: package-res/resources/web/pentaho/type/changes/Changeset.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+     *
+     * @returns {boolean} `true` if at least one property change is defined, `false` otherwise.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: suggested wording:

'true' is at least one property change is defined, 'false' if no property changes are defined.

File: package-res/resources/web/pentaho/type/changes/Changeset.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+     * @return {pentaho.type.Value} The original value of the property (before the change).
+     */
+    getOld: function(key) {
+      if(!this.has(key)) return null;
+
+      return this.getChange(key).oldValue;
+    },
+
+    /**
+     * Updates the owning [complex]{@linkplain pentaho.type.Complex} with this changes in this changeset.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: suggest:
Updates the owning [complex]{@linkplain pentaho.type.Complex} with the changes in this changeset.

File: package-res/resources/web/pentaho/type/changes/Changeset.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+     */
+    _freeze: function() {
+      O.eachOwn(this._properties, function(change) {
+        if(change) Object.freeze(change);
+      });
+      Object.freeze(this._properties);
+    },
+
+    /**
+     * Updates the owning [complex]{@linkplain pentaho.type.Complex} with this changes in this changeset.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: suggest:

Updates the owning [complex]{@linkplain pentaho.type.Complex} with the changes in this changeset.

File: package-res/resources/web/pentaho/type/changes/ListChange.js [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    get newValue() {
+      var newValue = this._newValue;
+      if(newValue) return newValue;
+
+      this._newValue = newValue = this.apply(this.oldValue.clone());
+      return newValue;
+    },
+
+    /**
+     * Adds, removes and/or updates elements to the element list.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/ListChange.js]: Adds, removes, and/or updates elements to the element list.

File: package-res/resources/web/pentaho/type/changes/ListChange.js [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    },
+
+    /**
+     * Creates an operation that updates an element in the list,
+     * and appends that operation to the list of changes.
+     * 
+     * In an `update` operation, the reference to the element does not change, but its contents does.
+     *
+     * @param {!pentaho.type.Element} elem - The object (already in the list) that will be updated.
+     * @param {!pentaho.type.Element} other - The object with the content will be used for updating the list.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/ListChange.js]: suggested wording:

The object with the content that will be used for updating the list.

File: package-res/resources/web/pentaho/type/changes/Update.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  
+  /**
+   * @name Update
+   * @memberOf pentaho.type.changes
+   * @class
+   * @extends pentaho.type.changes.Operation
+   * @amd pentaho/type/changes/Update
+   *
+   * @classDesc Describes an operation that updates an element in a list.
+   * In an `update` operation, the reference to the element does not change, but its contents does.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Update.js]: suggesting wording:

In an `update` operation, the reference to the element does not change, but its content does.

File: package-res/resources/web/pentaho/type/changes/ListChange.js [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+    _removeOne: function(elem, key) {
+      this._removeKeys[key] = elem;
+      this.changes.push(new RemoveOne(elem, key));
+    },
+
+    /**
+     * Creates an operation that updates an element in the list,
+     * and appends that operation to the list of changes.
+     * 
+     * In an `update` operation, the reference to the element does not change, but its contents does.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/ListChange.js]: suggested wording:

In an `update` operation, the reference to the element does not change, but its content does.

Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/ListChange.js]: ✅

File: package-res/resources/web/pentaho/type/changes/ListChange.js (hunk 10 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/ListChange.js]: ✅

File: package-res/resources/web/pentaho/type/changes/Changeset.js (hunk 4 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: ✅

File: package-res/resources/web/pentaho/type/changes/Changeset.js (hunk 5 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: ✅

File: package-res/resources/web/pentaho/type/changes/Changeset.js (hunk 6 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: ✅

File: package-res/resources/web/pentaho/type/changes/Changeset.js (hunk 8 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: ✅

File: package-res/resources/web/pentaho/type/changes/ListChange.js (hunk 9 above)
Comment (carlosrusso) [inline package-res/resources/web/pentaho/type/changes/ListChange.js]: ✅

File: package-res/resources/web/pentaho/type/changes/Add.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+   * @class
+   * @extends pentaho.type.changes.Operation
+   * @amd pentaho/type/changes/Add
+   *
+   * @classDesc Describes an operation that adds an element to a list.
+   *
+   * @constructor
+   * @description Creates an instance.
+   *
+   * @param {!pentaho.type.Element} elem - The object to be added to the list.
```
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Add.js]: just a test to show Joe how to do these comments :)

File: package-res/resources/web/pentaho/type/changes/Changeset.js (hunk 6 above)
Comment (swagner8412) [inline package-res/resources/web/pentaho/type/changes/Changeset.js]: @carlosrusso i made a typo...oopsie!

'true' if at least one property change is defined, 'false' if no property changes are defined.

**Comment by nantunes** [comment]
#748 seems to be the final branch, closing this PR.

**Pull request closed without merging**